### PR TITLE
ccm arraybuffer implementation fix

### PR DESCRIPTION
**Pull request by miha-stopar**
There was a ccm.encrypt call in sjcl._encrypt in convenience.js:
 p.ct = sjcl.arrayBuffer.ccm.encrypt(prp, plaintext, p.iv, adata, p.ts);

I think this is wrong for two reasons:
1) ccm.compat_encrypt is meant as a drop in replacement for old (non ArrayBuffers) ccm.encrypt (and the additional conditional statement checking whether the plaintext is ArrayBuffer can be this way removed)
2) ccm.encrypt returns ArrayBuffer which is not properly encoded later in sjcl.encrypt function (and leaves the ct value as an empty string) and obviously making decryption impossible

I modified compat_encrypt in the way that it accepts either bitArray or ArrayBuffer (and there is an additional test with plaintext being pushed into encryption as an ArrayBuffer).

Also, two minor points: I guess checking for iv having sufficient length has been mistakenly deleted and padding_count should be defaulted to 16, not 8?

### CODE REVIEW

**Comment by Nilos** [comment]
The tests are red :(
I like the idea in general. In the beginning I wanted to keep the data going through array buffer low (so developers have to very specifically decide to use it) but I guess we can move a step forward (this change would still require manual activation of the arrayBuffer module).

Please fix tests, will add more comments regarding the code specifics then.

**Comment by miha-stopar** [comment]
Not sure how this happens, but somehow the codec.arrayBuffer gets like this on Travis:

{ fromBits: [Function],
  toBits: [Function],
  padBuffer: [Function],
  Fa: [Function],
  Ea: [Function] }

Somehow toBuffer and hexDumpBuffer functions are renamed into Fa and Ea. Any idea?

**Comment by Nilos** [comment]
Looks like the closure compiler is not working correctly.
The whole build pipeline is really old but I did not find the time yet to fix it.

**Comment by Nilos** [comment]
Try adding usage comments (like for fromBits) to the functions.

File: core/codecArrayBuffer.js [hunk 1]
```diff
@@ -88,7 +88,36 @@ sjcl.codec.arrayBuffer = {
...
+      ol += padding_count - (ol%padding_count);
+    }
+
+    var out = new Uint8Array(ol);
+      out.set(new Uint8Array(buffer), 0);
+      return out.buffer;
+  },
+
+  /** Ciphertext is stored in sjcl encoded bitArray - this function converts it back to buffer **/
+  toBuffer: function (bitArray) {
```
Comment (Nilos) [inline core/codecArrayBuffer.js]: I am still not sure why you need this. Shouldn't the fromBits function do almost the same?

**Comment by miha-stopar** [comment]
Yes, it is very similar, but consider the case when you encrypt some file:

1) Load the file content into Uint8Array, for example with a file with 5 bytes you get something like:
 [49, 50, 51, 52, 10]

2. Encrypt the file and get something like: 

"{"iv":"5S6N11sXkStqChk/tfopzHA==","v":1,"iter":1000,"ks":128,"ts":64,"mode":"ccm","adata":"","cipher":"aes","ct":"geQzvprrZg3+fP8NxA=="}"

3. Decrypt the file and you get something like:

[825373492, 8796260794368]

4. If you check the number of bytes of this bitArray (let it be called d), you get 5:

sjcl.bitArray.bitLength(d) / 8 = 5

5. Using toBuffer function can be obtained the original file content:

t = new Uint8Array(sjcl.codec.arrayBuffer.toBuffer(d));

where t = [49, 50, 51, 52, 10]

6. However, using fromBits:

t1 = new Uint8Array(sjcl.codec.arrayBuffer.fromBits(d))

where t1 = [49, 50, 51, 52, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

But yes, I can remove toBuffer and we can simply leave this slicing to be done by users of the library. 
BTW I added comments to both functions.

**Comment by Nilos** [comment]
Couldn't you just use fromBits with no padding? (`sjcl.codec.arrayBuffer.fromBits(d, false)`)

File: core/ccmArrayBuffer.js [hunk 2]
```diff
@@ -94,7 +97,11 @@ sjcl.arrayBuffer.ccm = {
     tlen = tlen || sjcl.arrayBuffer.ccm.defaults.tlen;
     ol = ol || plaintext_buffer.byteLength;
     tlen = Math.ceil(tlen/8);
-    
+
+    if (ivl < 7) {
```
Comment (Nilos) [inline core/ccmArrayBuffer.js]: Well spotted!

File: core/codecArrayBuffer.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ sjcl.codec.arrayBuffer = {
   fromBits: function (arr, padding, padding_count) {
     var out, i, ol, tmp, smallest;
     padding = padding==undefined  ? true : padding
-    padding_count = padding_count || 8
+    padding_count = padding_count || 16
```
Comment (Nilos) [inline core/codecArrayBuffer.js]: Why do you prefer 16?

File: core/ccmArrayBuffer.js [hunk 4]
```diff
@@ -34,11 +34,14 @@ sjcl.arrayBuffer.ccm = {
...
-        encrypted_obj,
-        ct,
-        tag;
+    var plaintext_buffer, ol, encrypted_obj, ct, tag;
+    if (plaintext instanceof ArrayBuffer) {
+      ol = plaintext.byteLength;
+      plaintext_buffer = sjcl.codec.arrayBuffer.padBuffer(plaintext);
+    } else {
+      ol = sjcl.bitArray.bitLength(plaintext)/8;
+      plaintext_buffer = sjcl.codec.arrayBuffer.fromBits(plaintext, true, 16);
```
Comment (Nilos) [inline core/ccmArrayBuffer.js]: If the default is true, 16 you can omit the values here.

File: core/codecArrayBuffer.js [hunk 5]
```diff
@@ -88,7 +88,41 @@ sjcl.codec.arrayBuffer = {
     return out;
   },
 
+  /** Add padding with zero bytes if needed.
+   * This is the same padding as applied in fromBits where
+   * bitArray gets converted to ArrayBuffer.*/
+  padBuffer: function (buffer, padding, padding_count) {
```
Comment (Nilos) [inline core/codecArrayBuffer.js]: Can we re-use this function in the fromBits function?

File: core/codecArrayBuffer.js [hunk 6]
```diff
@@ -88,7 +88,41 @@ sjcl.codec.arrayBuffer = {
...
+    var out = new Uint8Array(ol);
+      out.set(new Uint8Array(buffer), 0);
+      return out.buffer;
+  },
+
+  /** Convert decrypted bitArray into buffer with original length.
+   * When plaintext is obtained by decryption, it is in the form of bitArray.
+   * If this bitArray would be converted using fromBits, there would be some
+   * zeros appended at the end (compared to the original ArrayBuffer prior encryption). **/
+  toBuffer: function (bitArray) {
```
Comment (Nilos) [inline core/codecArrayBuffer.js]: Still not sure what the difference between this function and fromBits(bitArray, false) is

File: core/convenience.js [hunk 7]
```diff
@@ -64,8 +64,8 @@
     rp.key = password;
 
     /* do the encryption */
-    if (p.mode === "ccm" && sjcl.arrayBuffer && sjcl.arrayBuffer.ccm && plaintext instanceof ArrayBuffer) {
-      p.ct = sjcl.arrayBuffer.ccm.encrypt(prp, plaintext, p.iv, adata, p.ts);
+    if (p.mode === "ccm" && sjcl.arrayBuffer && sjcl.arrayBuffer.ccm) {
+      p.ct = sjcl.arrayBuffer.ccm.compat_encrypt(prp, plaintext, p.iv, adata, p.ts);
```
Comment (Nilos) [inline core/convenience.js]: This would enable arrayBuffer ccm for all encryption operations if the module is active, correct?

File: test/test.js [hunk 8]
```diff
@@ -132,3 +132,18 @@ sjcl.codec.hex = sjcl.codec.hex ||
     return sjcl.bitArray.clamp(out, len*4);
   }
 };
+
+// CCM ArrayBuffer implementation test relies on sjcl.codec.arrayBuffer.toBuffer
+sjcl.codec.arrayBuffer = sjcl.codec.arrayBuffer ||
```
Comment (Nilos) [inline test/test.js]: I guess this is a "hotfix" to make the tests work?
In test environments the code should always be built with all modules!

File: core/convenience.js (hunk 7 above)
Comment (Nilos) [inline core/convenience.js]: We might want to at least give the library user the option to have sjcl.arrayBuffer enabled but still not use it in the convenience function. Maybe a params attribute?

File: core/codecArrayBuffer.js (hunk 3 above)
Comment (miha-stopar) [inline core/codecArrayBuffer.js]: It throws an exception when set to 8, see 
mac[2] ^= data.getUint32(i+8);

in _computeTag

File: core/ccmArrayBuffer.js (hunk 4 above)
Comment (miha-stopar) [inline core/ccmArrayBuffer.js]: Done.

File: core/codecArrayBuffer.js (hunk 5 above)
Comment (miha-stopar) [inline core/codecArrayBuffer.js]: Done.

File: core/codecArrayBuffer.js (hunk 6 above)
Comment (miha-stopar) [inline core/codecArrayBuffer.js]: Yes, somehow I missed it. It is fixed now - function fromBits is used a test and toBuffer is removed.
However, I don't know what to do to fix a build failure (now padBuffer is not found).

File: test/test.js (hunk 8 above)
Comment (miha-stopar) [inline test/test.js]: Yes, I agree. I was just trying to see what is the problem. I have removed it now.

File: core/convenience.js (hunk 7 above)
Comment (miha-stopar) [inline core/convenience.js]: Which operations do you mean? Why not use it in the convenience function? If not, library user would have to take care of the serialization of the ciphertext and all other data (iv, ts...) by himself.

**Comment by Nilos** [comment]
The tests are still red :( I will close this PR for now and we can re-open it if the tests are green.
Run the tests locally:
```
./configure --with-all
npm test
```

**Comment by miha-stopar** [comment]
Ok, now I see it (I was testing it without compression) - I have removed the description for padBuffer function and it works now. Can you reopen it?

**Comment by Nilos** [comment]
It is still failing in node 0.8 (most likely due to a bug in the arraybuffer implementation in that node)
It would still be nice if we had a check to fallback to non-arraybuffer in broken versions

**Comment by miha-stopar** [comment]
I added a check and an alternative approach for versions when slice method is not available - seems to be ok now.

File: core/convenience.js [hunk 9]
```diff
@@ -131,12 +131,12 @@
...
     /* do the decryption */
-    if (p.mode === "ccm" && sjcl.arrayBuffer && sjcl.arrayBuffer.ccm && p.ct instanceof ArrayBuffer) {
-      ct = sjcl.arrayBuffer.ccm.decrypt(prp, p.ct, p.iv, p.tag, adata, p.ts);
+    if (p.mode === "ccm" && sjcl.arrayBuffer && sjcl.arrayBuffer.ccm) {
+      ct = sjcl.arrayBuffer.ccm.compat_decrypt(prp, p.ct, p.iv, adata, p.ts);
     } else {
       ct = sjcl.mode[p.mode].decrypt(prp, p.ct, p.iv, adata, p.ts);
     }
-
+
```
Comment (Nilos) [inline core/convenience.js]: unnecessary whitespace

File: core/codecArrayBuffer.js [hunk 10]
```diff
@@ -17,48 +18,48 @@ sjcl.codec.arrayBuffer = {
...
-
-    smallest = tmp.byteLength < out.byteLength ? tmp.byteLength : out.byteLength;
-    for(i=0; i<smallest; i++){
-      out.setUint8(i,tmp.getUint8(i));
+    var buffer;
+    if(tmp.buffer.hasOwnProperty("slice")){
+        buffer = tmp.buffer.slice(0, sjcl.bitArray.bitLength(arr)/8);
+    } else {
+	// for node 0.8
+	var n = new Uint8Array(tmp.buffer);
```
Comment (Nilos) [inline core/codecArrayBuffer.js]: whitespace is off here!

**Comment by miha-stopar** [comment]
Fixed.

File: core/convenience.js (hunk 7 above)
Comment (Nilos) [inline core/convenience.js]: There is one disadvantage in this implementation in comparison to the previous way:
If an ArrayBuffer is used as input, we still return ct as a sjcl-bitArray, making it necessary to transform the ct arrayBuffer into bitArray which requires unnecessary amounts of speed/time.

Maybe we can still return an arrayBuffer if we get an arrayBuffer as input?

Also because of the padding the behaviour for arrayBuffer input also changed. I'm still not sure why the old behaviour was not valid though (why do we need padding if our input is an arrayBuffer?)

From my point of view we should not use `compat_encrypt` / `compat_decrypt` if we get passed an arrayBuffer but simply `arrayBuffer.ccm.encrypt` / `decrypt`.

Comment (miha-stopar) [inline core/convenience.js]: Padding of ArrayBuffer is needed otherwise we get:

        mac[0] ^= data.getUint32(i);
                       ^
	Error: Index out of range.
    		at Object.sjcl.arrayBuffer.ccm._computeTag (evalmachine.<anonymous>:4952:24)
    		at Object.sjcl.arrayBuffer.ccm.encrypt (evalmachine.<anonymous>:4865:32)

In the first version padding was applied in compat_encrypt, however the encryption failed when encrypt was used directly (but there was no test for it).
 
To summarize what was now done:
 - CCM ArrayBuffer encryption now accepts only ArrayBuffer input
 - I removed compat_encrypt and compat_decrypt which were meant for encryption/decryption when input is a bitArray - now this is not needed as sjcl.codec.arrayBuffer.fromBits was fixed (removed padBuffer at the end - padBuffer is now called directly in encrypt and decrypt functions) and bitArray can be easily encrypted using CCM ArrayBuffer encryption simply by converting into ArrayBuffer prior the encryption (see tests):
	var plaintextArrayBuffer = sjcl.codec.arrayBuffer.fromBits(plaintextBits);
  	var ciphertext = sjcl.encrypt(sessionKey, plaintextArrayBuffer, params);
 - encode and decode in convenience.js have been modified to encode/decode ArrayBuffer into Base64, for this reason fromBase64 and toBase64 have been added to sjcl.codec.arrayBuffer (in the previous version the ArrayBuffer was converted into bitArray before encoding because encode didn't support ArrayBuffers)
 - related to the previous point: encoding of ArrayBuffer provides two properties (ciphertextBuffer and ciphertextTag) instead of only one (ct) - this way decoding knows whether to decode into ArrayBuffer or bitArray (sjcl.decrypt output is thus the same as sjcl.encrypt input)

**Comment by miha-stopar** [comment]
Any further comments here? I would like to finalize this PR if possible.

**Comment by Nilos** [comment]
I still don't like the PR.
Why does it return a base64 string if I put in an arraybuffer (why not an arraybuffer?). Why does it use differently named fields for that?
base64 is still way more expensive and arraybuffers can be transfered to a server using blob transfer (via http or websocket).
So from my point of view the behaviour should be to return same kind of type as is passed into the function (bitArray, string, arraybuffer

**Comment by miha-stopar** [comment]
Encryption takes arraybuffer and returns an object containing arraybuffer (the actual ciphertext) and bitArray (ciphertext's mac). Base64 is used only for serialization.
